sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -220,88 +233,151 @@ class BulkWriter(container: CosmosAsyncContainer,
+            var numberOfIntervalsWithIdenticalActiveOperationSnapshots = 0
+            var activeTasksSnapshot = activeTasks.get()
+            while (activeTasksSnapshot > 0 && errorCaptureFirstException.get == null) {
+              log.logInfo(
+                s"Waiting for pending activeTasks $activeTasksSnapshot, Context: ${operationContext.toString}")
+              val activeOperationsSnapshot = activeOperations.clone()
+              val awaitCompleted = pendingTasksCompleted.await(1, TimeUnit.MINUTES)
+              if (!awaitCompleted) {
+                val sb = new StringBuilder()
+                //scalastyle:off magic.number
+                val maxItemOperationsToShowInErrorMessage = 10
+                //scalastyle:on magic.number
+                activeOperationsSnapshot
+                  .take(maxItemOperationsToShowInErrorMessage)
+                  .foreach(itemOperation => {
+                    if (sb.nonEmpty) {
+                      sb.append(", ")
+                    }
+
+                    sb.append(itemOperation.getOperationType)
+                    sb.append("->")
+                    val ctx = itemOperation.getContext[OperationContext]
+                    sb.append(s"${ctx.partitionKeyValue}/${ctx.itemId}/${ctx.eTag}(${ctx.attemptNumber})")
+                  })
+
+                if (activeOperationsSnapshot.equals(activeOperations)) {
+                  numberOfIntervalsWithIdenticalActiveOperationSnapshots += 1
+                  log.logWarning(
+                    s"FlushAndClose has been waiting ${numberOfIntervalsWithIdenticalActiveOperationSnapshots} " +
+                      s"times for identical set of operations: ${sb} Context: ${operationContext.toString}"
+                  )
+                } else {
+                  numberOfIntervalsWithIdenticalActiveOperationSnapshots = 0
+                  log.logInfo(
+                    s"FlushAndClose is waiting for active operations: ${sb} Context: ${operationContext.toString}"
+                  )
+                }
 
-        // error handling, if there is any error and the subscription is cancelled
-        // the remaining tasks will not be processed hence we never reach activeTasks = 0
-        // once we do error handling we should think how to cover the scenario.
-        lock.lock()
-        try {
-          var activeTasksSnapshot = activeTasks.get()
-          while (activeTasksSnapshot > 0 || errorCaptureFirstException.get != null) {
-            log.logDebug(s"Waiting for pending activeTasks $activeTasksSnapshot, Context: ${operationContext.toString}")
-            pendingTasksCompleted.await()
-            activeTasksSnapshot = activeTasks.get()
-            log.logDebug(s"Waiting completed for pending activeTasks $activeTasksSnapshot, Context: ${operationContext.toString}")
+                if (numberOfIntervalsWithIdenticalActiveOperationSnapshots >= 15) {
+                  errorCaptureFirstException.set(new IllegalStateException(
```

Comment:
compareAndSet(null, xxxx), I see we have several different places will touch this exception

Reply:
Resolved in next iteration